### PR TITLE
TF generation fixes

### PR DESCRIPTION
**Pull request by Rocketknight1**
This makes a few small bugfixes to the TF code in the course, and updates the generation sections to use XLA (cc @gante). It also fixes a few bugs, uses `prepare_tf_dataset` where possible, and also fixes a strange issue in section 5 that I think broke it for both TF and PT (cc @sgugger). 

@lbourdois @lewtun sorry for the delay!

### CODE REVIEW

File: chapters/en/chapter7/5.mdx [hunk 1]
```diff
@@ -289,9 +289,10 @@ def preprocess_function(examples):
         max_length=max_input_length,
         truncation=True,
     )
-    labels = tokenizer(text_target=targets, max_length=max_target_length, truncation=True)
+    labels = tokenizer(
+        examples["review_title"], max_length=max_target_length, truncation=True
+    )
```
Comment (Rocketknight1) [inline chapters/en/chapter7/5.mdx:294]: `targets` wasn't defined, so I think this code just didn't work before? Not sure how it got like this.

File: chapters/en/chapter7/3.mdx [hunk 2]
```diff
@@ -636,18 +635,18 @@ in your favorite terminal and log in there.
 
 {#if fw === 'tf'}
 
-Once we're logged in, we can create our `tf.data` datasets. We'll just use the standard data collator here, but you can also try the whole word masking collator and compare the results as an exercise:
+Once we're logged in, we can create our `tf.data` datasets. We use `model.prepare_tf_dataset()` here which uses our model to automatically infer which columns should go into the dataset. If you want to control exactly what columns to use, you can use `dataset.to_tf_dataset()` instead. We'll just use the standard data collator here, but you can also try the whole word masking collator and compare the results as an exercise:
```
Comment (lewtun) [inline chapters/en/chapter7/3.mdx:638]:
```suggestion
Once we're logged in, we can create our `tf.data` datasets. To do so, we'll use the `prepare_tf_dataset()` method, which uses our model to automatically infer which columns should go into the dataset. If you want to control exactly which columns to use, you can use the `Dataset.to_tf_dataset()` method instead. To keep things simple, we'll just use the standard data collator here, but you can also try the whole word masking collator and compare the results as an exercise:
```

File: chapters/en/chapter7/4.mdx [hunk 3]
```diff
@@ -495,28 +495,42 @@ The score can go from 0 to 100, and higher is better.
 
 {#if fw === 'tf'}
 
-To get from the model outputs to texts the metric can use, we will use the `tokenizer.batch_decode()` method. We just have to clean up all the `-100`s in the labels; the tokenizer will automatically do the same for the padding token. Let's define a function that takes our model and a dataset and computes metrics on it. Because generation of long sequences can be slow, we subsample the validation set to make sure this doesn't take forever:
+To get from the model outputs to texts the metric can use, we will use the `tokenizer.batch_decode()` method. We just have to clean up all the `-100`s in the labels; the tokenizer will automatically do the same for the padding token. Let's define a function that takes our model and a dataset and computes metrics on it. We're also going to use a trick that hugely increases performance here - [compiling our generation code with XLA](https://huggingface.co/blog/tf-xla-generate). This works a lot better when our input shapes don't vary too much, so we pad them to multiples of 128, and make a new dataset with that padding collator.
```
Comment (lewtun) [inline chapters/en/chapter7/4.mdx:498]:
```suggestion
To get from the model outputs to texts the metric can use, we will use the `tokenizer.batch_decode()` method. We just have to clean up all the `-100`s in the labels; the tokenizer will automatically do the same for the padding token. Let's define a function that takes our model and a dataset and computes metrics on it. We're also going to use a trick that dramatically increases performance - compiling our generation code with [XLA](https://www.tensorflow.org/xla), TensorFlow's accelerated linear algebra compiler. XLA applies various optimizations to the model's computation graph, and results in significant improvements to speed and memory usage. As described in the Hugging Face [blog](https://huggingface.co/blog/tf-xla-generate), XLA works best when our input shapes don't vary too much. To handle this,  we'll pad our inputs to multiples of 128, and make a new dataset with the padding collator:
```

File: chapters/en/chapter7/4.mdx [hunk 4]
```diff
@@ -495,28 +495,42 @@ The score can go from 0 to 100, and higher is better.
...
+
+tf_generate_dataset = model.prepare_tf_dataset(
+    tokenized_datasets["validation"],
+    collate_fn=generation_data_collator,
+    shuffle=False,
+    batch_size=8,
+)
+
+
+@tf.function(jit_compile=True)
```
Comment (lewtun) [inline chapters/en/chapter7/4.mdx:517]: It would be great to have a sentence or two explaining this decorator

File: chapters/en/chapter7/5.mdx [hunk 5]
```diff
@@ -727,18 +728,41 @@ model.fit(
 )
 ```
 
-We got some loss values during training, but really we'd like to see the ROUGE metrics we computed earlier. To get those metrics, we'll need to generate outputs from the model and convert them to strings. Let's build some lists of labels and predictions for the ROUGE metric to compare (note that if you get import errors for this section, you may need to`!pip install tqdm`):
+We got some loss values during training, but really we'd like to see the ROUGE metrics we computed earlier. To get those metrics, we'll need to generate outputs from the model and convert them to strings. Let's build some lists of labels and predictions for the ROUGE metric to compare (note that if you get import errors for this section, you may need to`!pip install tqdm`). We're also going to use a trick that hugely increases performance here - [compiling our generation code with XLA](https://huggingface.co/blog/tf-xla-generate). This works a lot better when our input shapes don't vary too much, so we pad them to a fixed size, and make a new dataset with that padding collator:
```
Comment (lewtun) [inline chapters/en/chapter7/5.mdx:731]: If you agree with my previous suggestion, we could copy-paste it here too

File: chapters/en/chapter7/5.mdx (hunk 1 above)
Comment (lewtun) [reply chapters/en/chapter7/5.mdx:294]: image

File: chapters/en/chapter7/5.mdx [hunk 6]
```diff
@@ -727,18 +728,41 @@ model.fit(
...
+        max_new_tokens=32,
+    )
+
+
 all_preds = []
 all_labels = []
-for batch in tqdm(tf_eval_dataset):
-    predictions = model.generate(**batch)
+for batch, labels in tqdm(tf_generate_dataset):
+    print(batch["input_ids"].shape)
```
Comment (lewtun) [inline chapters/en/chapter7/5.mdx:762]: Delete?

File: chapters/en/chapter7/6.mdx [hunk 7]
```diff
@@ -515,7 +515,7 @@ model.fit(tf_train_dataset, validation_data=tf_eval_dataset, callbacks=[callback
 
 {:else}
 
-💡 If you have access to a machine with multiple GPUs, you can try using a `MirroredStrategy` context to substantially speed up training. You'll need to create a `tf.distribute.MirroredStrategy` object, and make sure that the `to_tf_dataset` commands as well as model creation and the call to `fit()` are all run in its `scope()` context. You can see documentation on this [here](https://www.tensorflow.org/guide/distributed_training#use_tfdistributestrategy_with_keras_modelfit).
+💡 If you have access to a machine with multiple GPUs, you can try using a `MirroredStrategy` context to substantially speed up training. You'll need to create a `tf.distribute.MirroredStrategy` object, and make sure that any `to_tf_dataset` or `prepare_tf_dataset` commands as well as model creation and the call to `fit()` are all run in its `scope()` context. You can see documentation on this [here](https://www.tensorflow.org/guide/distributed_training#use_tfdistributestrategy_with_keras_modelfit).
```
Comment (lewtun) [inline chapters/en/chapter7/6.mdx:518]:
```suggestion
💡 If you have access to a machine with multiple GPUs, you can try using a `MirroredStrategy` context to substantially speed up training. You'll need to create a `tf.distribute.MirroredStrategy` object, and make sure that any `to_tf_dataset()` or `prepare_tf_dataset()` methods as well as model creation and the call to `fit()` are all run in its `scope()` context. You can see documentation on this [here](https://www.tensorflow.org/guide/distributed_training#use_tfdistributestrategy_with_keras_modelfit).
```

**Review by lewtun — COMMENTED** [review]
Thanks a lot for fixing these sections @Rocketknight1 🔥 !

I've left some nits, but overall this looks sweeeeet

File: chapters/en/chapter7/5.mdx (hunk 1 above)
Comment (Rocketknight1) [reply chapters/en/chapter7/5.mdx:294]: That code was used for PyTorch too!

File: chapters/en/chapter7/4.mdx (hunk 4 above)
Comment (Rocketknight1) [reply chapters/en/chapter7/4.mdx:517]: Done!

File: chapters/en/chapter7/5.mdx (hunk 5 above)
Comment (Rocketknight1) [reply chapters/en/chapter7/5.mdx]: Done!

File: chapters/en/chapter7/5.mdx (hunk 6 above)
Comment (Rocketknight1) [reply chapters/en/chapter7/5.mdx]: Yep, sorry! Debug line got left behind.

**Comment by Rocketknight1** [comment]
@lewtun all fixes are in! Thanks for the really solid review.

File: chapters/en/chapter7/5.mdx (hunk 1 above)
Comment (lbourdois) [reply chapters/en/chapter7/5.mdx:294]: I confirm, I had to modify this line in the French adaptation for it to work

**Comment by Rocketknight1** [comment]
@lewtun are you okay to merge now?

**Review by lewtun — APPROVED** [review]
Thanks for iterating @Rocketknight1 - LGTM 🔥 !
